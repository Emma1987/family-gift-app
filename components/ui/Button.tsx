import { StyleSheet, TouchableOpacity, ViewProps } from 'react-native';

import { ThemedText } from '@/components/ThemedText';
import { ThemedView } from '@/components/ThemedView';
import { useThemeColor } from '@/hooks/useThemeColor';

type ButtonProps = ViewProps & {
    text: string;
    accessibilityLabel: string;
    size?: 'small' | 'normal';
    onButtonPress: () => void;
};

export function Button({ style = {}, text, accessibilityLabel, size = 'normal', onButtonPress }: ButtonProps) {
    const buttonBackgroundColor = useThemeColor({}, 'buttonBackgroundColor');
    const buttonTextColor = useThemeColor({}, 'buttonTextColor');

    const buttonStyle = [
        styles.button,
        size === 'normal' ? styles.normalButton : styles.smallButton,
        { backgroundColor: buttonBackgroundColor },
        style,
    ];

    return (
        <TouchableOpacity onPress={onButtonPress} accessibilityLabel={accessibilityLabel} accessibilityRole="button">
            <ThemedView style={buttonStyle} testID="button-inner-container">
                <ThemedText style={[styles.text, { color: buttonTextColor }]}>{text}</ThemedText>
            </ThemedView>
        </TouchableOpacity>
    );
}

const styles = StyleSheet.create({
    button: {
        borderRadius: 8,
    },
    normalButton: {
        paddingVertical: 10,
        paddingHorizontal: 20,
    },
    smallButton: {
        paddingVertical: 5,
        paddingHorizontal: 10,
    },
    text: {
        fontWeight: 'bold',
        textAlign: 'center',
    },
});
