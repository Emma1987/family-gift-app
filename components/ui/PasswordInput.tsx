import { useState } from 'react';
import { StyleSheet, TextInput as RNTextInput, TouchableOpacity, View } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { Controller, useFormContext, RegisterOptions } from 'react-hook-form';

import { useThemeColor } from '@/hooks/useThemeColor';
import { ThemedText } from '@/components/ThemedText';

type PasswordInputProps = {
    name: string;
    control: any;
    placeholder: string;
    accessibilityLabel: string;
    rules?: RegisterOptions;
    autocomplete?: 'password' | 'password-new';
};

export function PasswordInput({
    name,
    control,
    placeholder,
    accessibilityLabel,
    rules = {},
    autocomplete = 'password',
}: PasswordInputProps) {
    const inputBorderColor = useThemeColor({}, 'inputBorderColor');
    const inputColor = useThemeColor({}, 'inputColor');
    const { trigger } = useFormContext();

    const [isValueDisplayed, setIsValueDisplayed] = useState<boolean>(false);

    return (
        <View>
            <Controller
                control={control}
                name={name}
                rules={rules}
                render={({ field: { onChange, onBlur, value }, fieldState: { error } }) => (
                    <>
                        <View
                            style={[
                                styles.inputContainer,
                                { borderColor: inputBorderColor },
                                error && styles.inputError,
                            ]}
                        >
                            <RNTextInput
                                style={[styles.input, { color: inputColor }]}
                                onChangeText={onChange}
                                onBlur={() => {
                                    onBlur();
                                    trigger(name);
                                }}
                                value={value}
                                placeholder={placeholder}
                                accessibilityLabel={accessibilityLabel}
                                autoCorrect={false}
                                secureTextEntry={!isValueDisplayed}
                                autoCapitalize="none"
                                autoComplete={autocomplete}
                            />
                            <TouchableOpacity style={styles.icon} onPress={() => setIsValueDisplayed((prev) => !prev)}>
                                <Ionicons
                                    name={isValueDisplayed ? 'eye-outline' : 'eye-off-outline'}
                                    size={25}
                                    color={inputColor}
                                    accessibilityLabel="Changer la visibilitÃ© du mot de passe"
                                    accessibilityRole="button"
                                    importantForAccessibility="no-hide-descendants"
                                />
                            </TouchableOpacity>
                        </View>
                        {error?.message && <ThemedText style={styles.errorText}>{error.message}</ThemedText>}
                    </>
                )}
            />
        </View>
    );
}

const styles = StyleSheet.create({
    inputContainer: {
        flexDirection: 'row',
        alignItems: 'center',
        borderBottomWidth: 1,
        marginBottom: 20,
    },
    input: {
        flex: 1,
        height: 45,
        fontWeight: '500',
    },
    icon: {
        padding: 10,
    },
    inputError: {
        borderBottomColor: 'red',
    },
    errorText: {
        color: 'red',
        fontSize: 12,
        marginTop: -20,
    },
});
