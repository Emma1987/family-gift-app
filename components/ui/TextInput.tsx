import { KeyboardTypeOptions, StyleSheet, TextInput as RNTextInput, View } from 'react-native';
import { Controller, useFormContext, RegisterOptions } from 'react-hook-form';

import { ThemedText } from '@/components/ThemedText';
import { useThemeColor } from '@/hooks/useThemeColor';

type TextInputProps = {
    name: string;
    control: any;
    placeholder: string;
    accessibilityLabel: string;
    rules?: RegisterOptions;
    keyboardType?: KeyboardTypeOptions;
};

export function TextInput({
    name,
    control,
    placeholder,
    accessibilityLabel,
    rules = {},
    keyboardType = 'default',
}: TextInputProps) {
    const inputBorderColor = useThemeColor({}, 'inputBorderColor');
    const inputColor = useThemeColor({}, 'inputColor');
    const { trigger } = useFormContext();

    return (
        <Controller
            control={control}
            name={name}
            rules={rules}
            render={({ field: { onChange, onBlur, value }, fieldState: { error } }) => (
                <View>
                    <RNTextInput
                        style={[
                            styles.input,
                            { borderColor: inputBorderColor, color: inputColor },
                            error && styles.inputError,
                        ]}
                        onChangeText={onChange}
                        onBlur={() => {
                            onBlur();
                            trigger(name);
                        }}
                        value={value}
                        placeholder={placeholder}
                        accessibilityLabel={accessibilityLabel}
                        keyboardType={keyboardType}
                        autoCorrect={false}
                        autoCapitalize="none"
                    />
                    {error?.message && <ThemedText style={styles.errorText}>{error.message}</ThemedText>}
                </View>
            )}
        />
    );
}

const styles = StyleSheet.create({
    input: {
        height: 45,
        borderBottomWidth: 1,
        fontWeight: 500,
        marginBottom: 20,
    },
    inputError: {
        borderBottomColor: 'red',
    },
    errorText: {
        color: 'red',
        fontSize: 12,
        marginTop: -20,
    },
});
