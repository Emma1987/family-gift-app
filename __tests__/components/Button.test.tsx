import { render, fireEvent, screen } from '@testing-library/react-native';
import { Button } from '@/components/ui/Button';

describe('Button Component', () => {
    it('renders the button correctly', () => {
        const { getByText, getByLabelText } = render(
            <Button text='Click me' accessibilityLabel='And see the magic!' onButtonPress={jest.fn()} />
        );

        expect(getByText('Click me')).toBeTruthy();
        expect(getByLabelText('And see the magic!')).toBeTruthy();
        const button = getByLabelText('And see the magic!');
        expect(button).toHaveProp('accessibilityRole', 'button');
    });

    it('calls onButtonPress when button is pressed', () => {
        const mockOnPress = jest.fn();

        const { getByLabelText } = render(
            <Button text='Click me' accessibilityLabel='And see the magic!' onButtonPress={mockOnPress} />
        );

        fireEvent.press(getByLabelText('And see the magic!'));
        expect(mockOnPress).toHaveBeenCalled();
    });

    it('applies normal size styles by default', () => {
        const { getByTestId } = render(
            <Button text='Click me' accessibilityLabel='And see the magic!' onButtonPress={jest.fn()} />
        );

        const button = getByTestId('button-inner-container');
        expect(button).toHaveStyle({
            paddingVertical: 10,
            paddingHorizontal: 20,
            borderRadius: 8,
        });
    });

    it('applies small size styles when size="small"', () => {
        const { getByTestId } = render(
            <Button text='Click me' accessibilityLabel='And see the magic!' onButtonPress={jest.fn()} size='small' />
        );

        const button = getByTestId('button-inner-container');
        expect(button).toHaveStyle({
            paddingVertical: 5,
            paddingHorizontal: 10,
            borderRadius: 8,
        });
    });

    it('matches the snapshot', () => {
        const tree = render(
            <Button text='Click me' accessibilityLabel='And see the magic!' onButtonPress={jest.fn()} />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
