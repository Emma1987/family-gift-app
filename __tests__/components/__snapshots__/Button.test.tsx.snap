// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches the snapshot 1`] = `
<View
  accessibilityLabel="And see the magic!"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#fffaf0",
        },
        [
          {
            "borderRadius": 8,
          },
          {
            "paddingHorizontal": 20,
            "paddingVertical": 10,
          },
          {
            "backgroundColor": "#151718",
          },
        ],
      ]
    }
    testID="button-inner-container"
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "fontWeight": "bold",
              "textAlign": "center",
            },
            {
              "color": "#f0e3c9",
            },
          ],
        ]
      }
    >
      Click me
    </Text>
  </View>
</View>
`;
